aggregator/src/aggregator/aggregation_job_continue.rs

```diff
@@ -1,5 +1,275 @@
+                    );
+                    aggregate_step_failure_counter.add(
+                        &Context::current(),
+                        1,
+                        &[KeyValue::new("type", "prepare_step_failure")],
+                    );
+                    response_prep_steps.push(PrepareStep::new(
+                        *prep_step.report_id(),
+                        PrepareStepResult::Failed(ReportShareError::VdafPrepError),
+                    ));
+                    ReportAggregationState::Failed(ReportShareError::VdafPrepError)
+                }
+            };
+
+            tx.update_report_aggregation(&report_aggregation.with_state(next_state))
+                .await?;
+        }
+
+        for report_agg in report_aggregations {
+            // This report was omitted by the leader because of a prior failure. Note that the
+            // report was dropped (if it's not already in an error state) and continue.
+            if matches!(report_agg.state(), ReportAggregationState::Waiting(_, _)) {
+                tx.update_report_aggregation(&report_agg.with_state(
+                    ReportAggregationState::Failed(ReportShareError::ReportDropped),
+                ))
+                .await?;
+            }
+        }
+
+        let helper_aggregation_job = helper_aggregation_job
+            // Advance the job to the leader's round
+            .with_round(leader_aggregation_job.round())
+            .with_state(match (saw_continue, saw_finish) {
+                (false, false) => AggregationJobState::Finished, // everything failed, or there were no reports
+                (true, false) => AggregationJobState::InProgress,
+                (false, true) => AggregationJobState::Finished,
+                (true, true) => {
+                    return Err(datastore::Error::User(
+                        Error::Internal(
+                            "VDAF took an inconsistent number of rounds to reach Finish state"
+                                .to_string(),
+                        )
+                        .into(),
+                    ))
+                }
+            });
+        tx.update_aggregation_job(&helper_aggregation_job).await?;
+
+        accumulator.flush_to_datastore(tx, vdaf).await?;
+
+        Ok(AggregationJobResp::new(response_prep_steps))
+    }
+
+    /// Fetch previously-computed prepare message shares and replay them back to the leader.
+    pub(super) fn replay_aggregation_job_round<C, const L: usize, Q, A>(
+        report_aggregations: Vec<ReportAggregation<L, A>>,
+    ) -> Result<AggregationJobResp, datastore::Error>
+    where
+        C: Clock,
+        Q: AccumulableQueryType,
+        A: vdaf::Aggregator<L, 16> + 'static + Send + Sync,
+        for<'a> A::PrepareState: Send + Sync + Encode + ParameterizedDecode<(&'a A, usize)>,
+    {
+        // DAP suggests that when detecting a replayed aggregation job continuation request, helper
```

Comment:
Storing a request hash is the approach I had planned to take from the get-go (which is why that's the approach suggested by the DAP text). I agree about the coherence problem, though what I came around to yesterday is that maybe it's not so bad to provide surprising responses to a leader that's already misbehaving, possibly with malicious intent.

I think what I'll do is take a swing at the request-hash based solution and see how complicated it gets (I think it may not be that bad), and then we can better weigh that against the coherence argument. In the interest of not further growing this PR, I'll go back to what I had a couple commits ago, which is a TODO and `#[ignore]` on the test that exercises this case.

Reply:
Turns out it was pretty easy to do: #1091